flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStorage.java

```diff
@@ -134,6 +149,37 @@ public CheckpointStorageLocation initializeLocationForCheckpoint(long checkpoint
 				writeBufferSize);
 	}
 
+	@Override
+	public void shutDown(JobStatus jobStatus) throws IOException {
```

Comment:
The logic to check whether checkpoint folders should be discarded are the same for Fs and MemoryBackend checkpoint storage, while the difference lies in what folders to discard. So we could extract them out into `AbstractFsCheckpointStorage` and introduce a new `doDiscard` abstract method for implementations to override.